### PR TITLE
Palm detection support

### PR DESCRIPTION
**Pull request by quarkw**
Got a chance to redo my changes with the swift refactor,  resolving #2698

This feature can help avoid inadvertent triggering of the fingers, but can also be used as an enhancement as it's own "key" to modulate behaviours and provide a new layer of keys.

Here's a demo with the IgnoredAreaView, including a new palmThreshold setting:

Palm demo

Modifies the multitouch extension to add the following palm counts

I left out the quarterArea counts but maybe those would be useful?

```
upperHalfAreaPalmCount
lowerHalfAreaPalmCount
leftHalfAreaPalmCount
rightHalfAreaPalmCount
totalPalmCount
```

I didn't rename any of the multitouch extension stuff, which refers to all touch interactions as "fingers", as that would make rebasing more difficult if this doesn't get merged in, or takes a while to get merged in.

Introduces a new `palmThreshold` setting to configure how large of an area needs to be detected to be considered as a palm.

All palm events should respect ignore areas, as well as the touch detection. Release detection does not apply, since palm release should go right into finger detection.

### CODE REVIEW

File: src/apps/MultitouchExtension/src/FingerManager.swift [hunk 1]
```diff
@@ -80,8 +90,9 @@ class FingerManager: ObservableObject {
     //
 
     for s in states {
-      if s.device == device && s.frame != frame && s.touchedPhysically {
+      if s.device == device && s.frame != frame && (s.touchedPhysically && !s.palmed) {
```
Comment (quarkw) [inline src/apps/MultitouchExtension/src/FingerManager.swift:93]: Looking at this logic here, I'm not sure I changed this expression correctly.

I don't really understand the purpose of this block to be honest. If the frame changes (from my testing, isn't frame just monotonically increasing?), we disable `s.touchedPhysically` and start the untouched delay task? Why is that?

File: src/apps/MultitouchExtension/src/Views/Settings/IgnoredAreaView.swift [hunk 2]
```diff
@@ -93,9 +94,13 @@ struct IgnoredAreaView: View {
...
 
           ForEach(fingerManager.states) { state in
-            if state.touchedPhysically || state.touchedFixed {
-              let diameter = 10.0
-              let color = state.ignored ? Color.black : Color.red
+            if state.touchedPhysically || state.touchedFixed || state.palmed {
+              let minDiameter = 5.0
+              let thresholdDiameter = 15 + palmThreshold * 5
+              let palmed = state.size > palmThreshold
+              let diameter = minDiameter + (thresholdDiameter - minDiameter) * (state.size/palmThreshold)
```
Comment (tekezo) [inline src/apps/MultitouchExtension/src/Views/Settings/IgnoredAreaView.swift:101]: Good UI updates 👍

**Comment by tekezo** [comment]
I'm updating logic in https://github.com/pqrs-org/Karabiner-Elements/pull/3564.

File: src/apps/MultitouchExtension/src/FingerManager.swift [hunk 3]
```diff
@@ -118,35 +129,51 @@ class FingerManager: ObservableObject {
         continue
       }
 
-      if !s.touchedFixed {
-        continue
-      }
+      if s.touchedFixed {
+        if s.touchedPhysically {
```
Comment (tekezo) [inline src/apps/MultitouchExtension/src/FingerManager.swift:133]: We need to increment the count even if touchedPhysically is false while touchedFixed is true.
This happens when the release detection delay is set.

Screenshot 2023-07-29 at 16 02 49

File: src/apps/MultitouchExtension/src/FingerManager.swift (hunk 1 above)
Comment (tekezo) [reply src/apps/MultitouchExtension/src/FingerManager.swift:93]: It is simpler to treat it as a simple flag in the same way as `ignored`.
I updated the logic in #3564.
https://github.com/pqrs-org/Karabiner-Elements/pull/3564/files?w=1

**Comment by tekezo** [comment]
@quarkw Thanks for a very nice feature!
I'll merge the changes with this changes https://github.com/pqrs-org/Karabiner-Elements/pull/3564.

As far as I used it, once it was recognised as palmed, I found it more convenient to have it recognised as palmed until the palm is released even if size < threshold, so I changed it that way.

https://github.com/pqrs-org/Karabiner-Elements/pull/3564/files#diff-a5eedcda1691d9e12007d48e0d09f9b3b81a7e75850718071267374dad9c14f0R70-R75

If the original behavior is more desirable for you, we can revert it by the following changes.
And I think we should add a option to swtich the behavior if you want to use the original behavior. So, let me know if new behavior is problematic.

```diff
-      if Double(finger.size) > palmThreshold {
-        s.palmed = true
-      }
+     s.palmed = Double(finger.size) > palmThreshold
```

**Comment by quarkw** [comment]
> @quarkw Thanks for a very nice feature! I'll merge the changes with this changes #3564.
> 
> As far as I used it, once it was recognised as palmed, I found it more convenient to have it recognised as palmed until the palm is released even if size < threshold, so I changed it that way.
> 
> https://github.com/pqrs-org/Karabiner-Elements/pull/3564/files#diff-a5eedcda1691d9e12007d48e0d09f9b3b81a7e75850718071267374dad9c14f0R70-R75
> 
> If the original behavior is more desirable for you, we can revert it by the following changes. And I think we should add a option to swtich the behavior if you want to use the original behavior. So, let me know if new behavior is problematic.
> 
> ```diff
> -      if Double(finger.size) > palmThreshold {
> -        s.palmed = true
> -      }
> +     s.palmed = Double(finger.size) > palmThreshold
> ```

No problem!

I had considered keeping the palm behavior once activated and agree it makes sense from a UX perspective.

I think configurability might clutter the settings. Disabling the feature can effectively be done by increasing the palm size to be very large. If we wanted this setting to be used this way,  we would probably want to increase the max value from 10 to something larger (we may want to increase the max from 10 anyway, for those with very large hands).

Also, do you think it makes sense to have the quarterArea variables for the palm counts? I only added the total and halfArea counts.

**Comment by tekezo** [comment]
@quarkw This feature is available since v14.12.9 🎉 
https://github.com/pqrs-org/Karabiner-Elements/releases/tag/beta

> No problem!

Thanks for the confirmation!
So, I keep this behavior, and I won't add settings.

> Also, do you think it makes sense to have the quarterArea variables for the palm counts? I only added the total and halfArea counts.

Unlike fingers, we cannot control the touch area with the palm finely, so I think halfArea is enough, too 👍

**Pull request merged**